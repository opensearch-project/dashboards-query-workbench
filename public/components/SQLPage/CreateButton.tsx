/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton, EuiComboBoxOptionOption, EuiContextMenu, EuiPopover } from '@elastic/eui';
import React, { useState } from 'react';
import {
  COVERING_INDEX_QUERY,
  CREATE_DATABASE_QUERY,
  CREATE_TABLE_QUERY,
  SKIPPING_INDEX_QUERY,
} from '../../../common/constants';

interface CreateButtonProps {
  updateSQLQueries: (query: string) => void;
  selectedDatasource: EuiComboBoxOptionOption[];
}

export const CreateButton = ({ updateSQLQueries, selectedDatasource }: CreateButtonProps) => {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  const [selectedOption, setSelectedOption] = useState(null);

  const closePopover = () => {
    setIsPopoverOpen(false);
  };

  const togglePopover = (option) => {
    setSelectedOption(option);
    setIsPopoverOpen(!isPopoverOpen);
  };

  const handleSubMenuClick = (query: string) => {
    updateSQLQueries(query);
    closePopover();
  };

  const fromSqlItems = [
    {
      name: 'Create Database',
      onClick: () => handleSubMenuClick(CREATE_DATABASE_QUERY),
    },
    {
      name: 'Create Table',
      onClick: () => handleSubMenuClick(CREATE_TABLE_QUERY),
    },
  ];

  const acceleratedIndexItems = [
    {
      name: 'Skipping Index',
      onClick: () => handleSubMenuClick(SKIPPING_INDEX_QUERY),
    },
    {
      name: 'Covering Index',
      onClick: () => handleSubMenuClick(COVERING_INDEX_QUERY),
    },
  ];

  const button = (
    <EuiButton iconType="arrowDown" iconSide="right" onClick={() => togglePopover(null)}>
      Create
    </EuiButton>
  );

  return (
    <>
      {selectedDatasource && selectedDatasource[0].label !== 'OpenSearch' && (
        <EuiPopover
          button={button}
          isOpen={isPopoverOpen}
          closePopover={closePopover}
          panelPaddingSize="s"
          anchorPosition="downLeft"
        >
          <EuiContextMenu
            initialPanelId={0}
            panels={[
              {
                id: 0,
                title: 'Create options',
                items: [
                  {
                    name: 'Spark SQL',
                    panel: 1,
                  },
                  {
                    name: 'Acceleration Index',
                    panel: 2,
                  },
                ],
              },
              {
                id: 1,
                title: 'SPARK SQL Options',
                items: fromSqlItems,
              },
              {
                id: 2,
                title: 'Acceleration Index Options',
                items: acceleratedIndexItems,
              },
            ]}
          />
        </EuiPopover>
      )}
    </>
  );
};
