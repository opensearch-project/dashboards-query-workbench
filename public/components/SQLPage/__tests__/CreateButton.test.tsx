/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { CreateButton } from '../CreateButton';

describe('CreateButton', () => {
  const mockUpdateSQLQueries = jest.fn();

  const mockSelectedDatasource = [{ label: 'glue_1' }];

  it('opens and closes the popover', async () => {
    render(
      <CreateButton
        updateSQLQueries={mockUpdateSQLQueries}
        selectedDatasource={mockSelectedDatasource}
      />
    );

    expect(screen.queryByText('Create Database')).not.toBeInTheDocument();

    fireEvent.click(screen.getByText('Create'));

    expect(screen.getByText('Spark SQL')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Spark SQL'));
    
    expect(document.body.children[0]).toMatchSnapshot();

   
    await waitFor(() => {
      expect(screen.queryByText('Create Database')).toBeInTheDocument();
    });
    expect(document.body.children[0]).toMatchSnapshot();

  });

  it('selects an option in the Spark SQL submenu', async () => {
    render(
      <CreateButton
        updateSQLQueries={mockUpdateSQLQueries}
        selectedDatasource={mockSelectedDatasource}
      />
    );

    fireEvent.click(screen.getByText('Create'));

    fireEvent.click(screen.getByText('Spark SQL'));

    await waitFor(() => {
      expect(screen.getByText('Create Database')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('Create Database'));

    expect(mockUpdateSQLQueries).toHaveBeenCalledWith(`CREATE DATABASE datasource.database`);
  });

});