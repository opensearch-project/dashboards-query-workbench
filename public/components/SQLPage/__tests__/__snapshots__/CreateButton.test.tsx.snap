// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateButton opens and closes the popover 1`] = `
<div>
  <div
    class="euiPopover euiPopover--anchorDownLeft"
  >
    <div
      class="euiPopover__anchor"
    >
      <button
        class="euiButton euiButton--primary"
        type="button"
      >
        <span
          class="euiButtonContent euiButtonContent--iconRight euiButton__content"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          />
          <span
            class="euiButton__text"
          >
            Create
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`CreateButton opens and closes the popover 2`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
>
  <div
    class="euiPopover euiPopover--anchorDownLeft euiPopover-isOpen"
  >
    <div
      class="euiPopover__anchor"
    >
      <button
        class="euiButton euiButton--primary"
        type="button"
      >
        <span
          class="euiButtonContent euiButtonContent--iconRight euiButton__content"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoaded euiButtonContent__icon"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.069 5.157 8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0Z"
              fill-rule="non-zero"
            />
          </svg>
          <span
            class="euiButton__text"
          >
            Create
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
